Workers react framework guide (#20214)
cloudflare/cloudflare-docs: src/content/docs/workers/frameworks/framework-guides/react.mdx

---
pcx_content_type: how-to
title: React
head: []
description: Create a React application and deploy it to Cloudflare Workers with Workers Assets.
---

import {
	Badge,
	Description,
	InlineBadge,
	Render,
	PackageManagers,
} from "~/components";

In this guide, you will create a new [React](https://react.dev/) application and deploy to Cloudflare Workers (with the new [<InlineBadge preset="beta" /> Workers Assets](/workers/static-assets/)).

## 1. Set up a new project

Use the [`create-cloudflare`](https://www.npmjs.com/package/create-cloudflare) CLI (C3) to set up a new project. C3 will create a new project directory, use code from the official React template, and provide the option to deploy instantly.

To use `create-cloudflare` to create a new React project with <InlineBadge preset="beta" /> Workers Assets, run the following command:

<PackageManagers
	type="create"
	pkg="cloudflare@latest my-react-app"

> 	args={"--framework=react --experimental"}

Reviewer: ```suggestion
	args={"--framework=react --platform=workers"}
```

(Once https://github.com/cloudflare/workers-sdk/pull/8371 is released.)

/>

<Render
	file="c3-post-run-steps"
	product="workers"
	params={{
		category: "web-framework",
		framework: "React",
	}}
/>

After setting up your project, change your directory by running the following command:

```sh
cd my-react-app
```

## 2. Develop locally

After you have created your project, run the following command in the project directory to start a local server. This will allow you to preview your project locally during development.

<PackageManagers type="run" args={"dev"} />

> ## 3. Deploy your Project

Reviewer 2: ```suggestion
## 3. Deploy your project
```

Your project can be deployed to a `*.workers.dev` subdomain or a [Custom Domain](/workers/configuration/routing/custom-domains/), from your own machine or from any CI/CD system, including [Cloudflare's own](/workers/ci-cd/builds/).

> The following command will build and deploy your project. If you're using CI, ensure you update your ["deploy command"](/workers/ci-cd/builds/configuration/#build-settings) configuration appropriately.

Reviewer 2: ```suggestion
The following command will build and deploy your project. If you are using CI, ensure you update your ["deploy command"](/workers/ci-cd/builds/configuration/#build-settings) configuration appropriately.
```

---
pcx_content_type: how-to
title: React
head: []
description: Create a React application and deploy it to Cloudflare Workers with Workers Assets.
---

import {
	Badge,
	Description,
	InlineBadge,
	Render,
	PackageManagers,
} from "~/components";

In this guide, you will create a new [React](https://react.dev/) application and deploy to Cloudflare Workers (with the new [<InlineBadge preset="beta" /> Workers Assets](/workers/static-assets/)).

## 1. Set up a new project

> Use the [`create-cloudflare`](https://www.npmjs.com/package/create-cloudflare) CLI (C3) to set up a new project. C3 will create a new project directory, initiate the React official setup tool, and provide the option to deploy instantly.

Reviewer 3: "initiate the React official setup tool" feels wrong to me.

I'd say "use code from the official React template" maybe?

> Use the [`create-cloudflare`](https://www.npmjs.com/package/create-cloudflare) CLI (C3) to set up a new project. C3 will create a new project directory, initiate the React official setup tool, and provide the option to deploy instantly.

Author: ```suggestion
Use the [`create-cloudflare`](https://www.npmjs.com/package/create-cloudflare) CLI (C3) to set up a new project. C3 will create a new project directory, use code from the official React template, and provide the option to deploy instantly.
```

<PackageManagers
	type="create"
	pkg="cloudflare@latest my-react-app"
	args={"--framework=react --experimental"}
/>

<Render
	file="c3-post-run-steps"
	product="workers"
	params={{
		category: "web-framework",
		framework: "React",
	}}
/>

After setting up your project, change your directory by running the following command:

```sh
cd my-react-app
```

## 2. Develop locally

After you have created your project, run the following command in the project directory to start a local server. This will allow you to preview your project locally during development.

<PackageManagers type="run" args={"dev"} />

## 3. Deploy your Project

Your project can be deployed to a `*.workers.dev` subdomain or a [Custom Domain](/workers/configuration/routing/custom-domains/), from your own machine or from any CI/CD system, including [Cloudflare's own](/workers/ci-cd/builds/).

The following command will build and deploy your project. If you're using CI, ensure you update your ["deploy command"](/workers/ci-cd/builds/configuration/#build-settings) configuration appropriately.

<PackageManagers type="run" args={"deploy"} />

---

> 

Author: ```suggestion
```

> ## Bindings

Author: ```suggestion
## Connect React with Workers bindings
```

> ## Connect React with Workers bindings

Author: ```suggestion
```

> 

Author: ```suggestion
```

> Your React application can be fully integrated with the Cloudflare Developer Platform, in both local development and in production, by using product bindings. The [React documentation](UPDATE) provides information about configuring bindings.

Author: @petebacondarwin Do you have a recommendation for what this link should be? (i.e. where it says "UPDATE") 
The other framework guides all link to the framework's documentation for cloudflare, or bindings, or runtime APIs but I couldn't find any of that in the React documentation. Is this where I should be mentioning Vite and/or linking to Vite documentation instead?

> Your React application can be fully integrated with the Cloudflare Developer Platform, in both local development and in production, by using product bindings. The [React documentation](UPDATE) provides information about configuring bindings.

Author: ```suggestion
Your React application cannot directly access Workers bindings, but it can interact with them through a Worker backend API. To do this, make [`fetch()` requests](/workers/runtime-apis/fetch/) from your React app to the Worker, which can then handle the request and connect to bindings. Learn how to [configure Workers bindings](/workers/runtime-apis/bindings/).
```

> Your React application cannot directly access Workers bindings, but it can interact with them through a Worker backend API. To do this, make [`fetch()` requests](/workers/runtime-apis/fetch/) from your React app to the Worker, which can then handle the request and connect to bindings. Learn how to [configure Workers bindings](/workers/runtime-apis/bindings/).

Author: @petebacondarwin
Thank you for explaining this, Pete! I wasn't sure how much of that explanation you intended for me to include here, so I wrote an abbreviated version. Let me know what you think.
(Sorry I probably should have done this as as suggested comment instead of a new commit :P )

> Your React application cannot directly access Workers bindings, but it can interact with them through a Worker backend API. To do this, make [`fetch()` requests](/workers/runtime-apis/fetch/) from your React app to the Worker, which can then handle the request and connect to bindings. Learn how to [configure Workers bindings](/workers/runtime-apis/bindings/).

Author: ```suggestion
```

## Static assets

> You can serve static assets in your React application by placing them in [the `./public/` directory](UPDATE). This can be useful for resource files such as images, stylesheets, fonts, and manifests.

Reviewer: For the public directory, this is a standard Vite setup. So I think we can just link to https://vite.dev/guide/assets

> You can serve static assets in your React application by placing them in [the `./public/` directory](UPDATE). This can be useful for resource files such as images, stylesheets, fonts, and manifests.

Author: ```suggestion
You can serve static assets in your React application by [placing them in the `./public/` directory](https://vite.dev/guide/assets#the-public-directory). This can be useful for resource files such as images, stylesheets, fonts, and manifests.
```
@petebacondarwin This ok?
